sql/src/test/java/io/crate/integrationtests/disruption/seqno/ConcurrentSeqNoVersioningIT.java

```diff
@@ -0,0 +1,756 @@
+            super("CAS-Update-" + threadNum);
+            this.roundBarrier = roundBarrier;
+            this.partitions = partitions;
+            this.timeoutSeconds = timeoutSeconds;
+            setDaemon(true);
+        }
+
+        public void run() {
+            while (stop == false) {
+                try {
+                    roundBarrier.await(70, TimeUnit.SECONDS);
+
+                    int numberOfUpdates = randomIntBetween(3, 13)  * partitions.size();
+                    for (int i = 0; i < numberOfUpdates; ++i) {
+                        final int keyIndex = random.nextInt(partitions.size());
+                        final Partition partition = partitions.get(keyIndex);
+
+                        final int seqNoChangePct = random.nextInt(100);
+
+                        // we use either the latest observed or the latest successful version, to increase chance of getting successful
+                        // CAS'es and races. If we were to use only the latest successful version, any CAS fail on own write would mean that
+                        // all future CAS'es would fail unless we guess the seqno/term below. On the other hand, using latest observed
+                        // version exclusively we risk a single CAS fail on a dirty read to cause the same. Doing both randomly and adding
+                        // variance to seqno/term should ensure we progress fine in most runs.
+                        Version version = random.nextBoolean() ? partition.latestObservedVersion() : partition.latestSuccessfulVersion();
+
+                        if (seqNoChangePct < 10) {
+                            version = version.nextSeqNo(random.nextInt(4) + 1);
+                        } else if (seqNoChangePct < 15) {
+                            version = version.previousSeqNo(random.nextInt(4) + 1);
+                        }
+
+                        final int termChangePct = random.nextInt(100);
+                        if (termChangePct < 5) {
+                            version = version.nextTerm();
+                        } else if (termChangePct < 10) {
+                            version = version.previousTerm();
+                        }
+
+                        Consumer<HistoryOutput> historyResponse = partition.invoke(version);
+                        try {
+                            execute("update t set x = ? where id = ? and _seq_no = ? and _primary_term = ?",
+                                    new Object[]{random.nextInt(), partition.id, version.seqNo, version.primaryTerm},
+                                    TimeValue.timeValueSeconds(timeoutSeconds));
+
+                            if (response.rowCount() > 0) {
+                                IndexResponseHistoryOutput historyOutput = new IndexResponseHistoryOutput(new Version(
+                                    version.primaryTerm,
+                                    version.seqNo + 1));
```

Comment:
`primary_term` and `seq_no` from the response must be used.
Also this should not be `version.seqNo + 1`, bit just the responded `seq_no`

Reply:
Great catch, thanks.